### PR TITLE
Add train task structure

### PR DESCRIPTION
**Pull request by guarin**
## Did you update [CHANGELOG.md](../CHANGELOG.md)?

- [ ] Yes
- [x] Not needed (internal change)

## Did you update the documentation?

- [ ] Yes
- [x] Not needed (internal change without effects for user)

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -112,6 +112,7 @@ warn_unused_ignores = false
 module = [
     "albumentations.*",
     "lightly.*",
+    "lightning_fabric.*",
```
Comment (stegmuel) [inline pyproject.toml:115]: With fabric we will be able to use torch 2.7? Can we have temporarily two distinct set of requirements?

File: src/lightly_train/_data/mask_semantic_segmentation_dataset.py [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+
+from lightly_train.types import MaskSemanticSegmentationDatasetItem
+
+
+class MaskSemanticSegmentationDataset(Dataset[MaskSemanticSegmentationDatasetItem]):
+    def __getitem__(self, index: int) -> MaskSemanticSegmentationDatasetItem:
+        return {
+            "image_filename": "abc.jpg",
+            "image": torch.rand(3, 224, 224, dtype=torch.uint8),
+            "mask": torch.rand(5, 224, 224, dtype=torch.uint8),
```
Comment (stegmuel) [inline src/lightly_train/_data/mask_semantic_segmentation_dataset.py:19]: If we go for semantic segmentation we should got for a single channel and values in the range [0, C-1]. I have tried stacking everything once and the overhead was huge.

Comment (stegmuel) [reply src/lightly_train/_data/mask_semantic_segmentation_dataset.py:19]: But we have to anticipate the case where some pixels could have multiple labels.

File: src/lightly_train/_task_models/dinov2_semantic_segmentation/dinov2_semantic_segmentation.py [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+from lightly_train._models.dinov2_vit.dinov2_vit_src.models.vision_transformer import (
+    DinoVisionTransformer,
+)
+from lightly_train._task_models.task_model import TaskModel
+
+
+class DINOv2SemanticSegmentation(TaskModel):
+    def __init__(self) -> None:
+        super().__init__()
+        self.backbone: DinoVisionTransformer
```
Comment (stegmuel) [inline src/lightly_train/_task_models/dinov2_semantic_segmentation/dinov2_semantic_segmentation.py:21]: Out of the box we will be able to load from `exported_last.pt`, right?

File: src/lightly_train/_data/mask_semantic_segmentation_dataset.py (hunk 2 above)
Comment (stegmuel) [reply src/lightly_train/_data/mask_semantic_segmentation_dataset.py:19]: Throwing some random info: very often it is needed to ignore some labels as they are the border of objects or background.

File: src/lightly_train/_task_models/task_model.py [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+# This source code is licensed under the license found in the
+# LICENSE file in the root directory of this source tree.
+#
+from torch import Module
+
+
+class TaskModel(Module):
+    """Base class for task-specific models that the user interacts with.
+
+    Must implement the forward method for inference. Must be pure PyTorch and not rely
```
Comment (stegmuel) [inline src/lightly_train/_task_models/task_model.py:14]: Why is the inference of the task-specific models not sufficient?

File: pyproject.toml (hunk 1 above)
Comment (guarin) [reply pyproject.toml:115]: This change here is only to ignore some mypy warnings. Fabric and Lightning are tied to each other with the same versions. We'll have the same torch 2.7 support as we have now. We could add extra requirements with `lightly-train[task]` but I don't think we should do it, it will make installing the library more complicated and is harder to maintain.

File: src/lightly_train/_data/mask_semantic_segmentation_dataset.py (hunk 2 above)
Comment (guarin) [reply src/lightly_train/_data/mask_semantic_segmentation_dataset.py:19]: Ah yes good point, I just put something there 😅 

> Throwing some random info: very often it is needed to ignore some labels as they are the border of objects or background.

This we can handle by adding more fields to `data`. We don't need this info in the batches itself.

File: src/lightly_train/_task_models/dinov2_semantic_segmentation/dinov2_semantic_segmentation.py (hunk 3 above)
Comment (guarin) [reply src/lightly_train/_task_models/dinov2_semantic_segmentation/dinov2_semantic_segmentation.py:21]: Yes that should work as long as we load it directly into the backbone with `self.backbone.load_state_dict`. For proper load/export on the model level we'll have to make some more changes.

File: src/lightly_train/_task_models/task_model.py (hunk 4 above)
Comment (guarin) [reply src/lightly_train/_task_models/task_model.py:14]: Sufficient for what?

Comment (stegmuel) [reply src/lightly_train/_task_models/task_model.py:14]: Answered myself.
